src/DynamoCore/Core/TrustedLocatationsManager.cs

```diff
@@ -0,0 +1,166 @@
+                var existing = TrustedLocations.FirstOrDefault(x => x.Equals(location));
+                if (settings != null && string.IsNullOrEmpty(existing))
+                {
+                    trustedLocations.Add(location);
+                    settings.TrustedLocations = trustedLocations;
+                    return true;
+                }
+            }
+            catch { }
+            return false;
+        }
+
+        /// <summary>
+        /// Remove a path from the Dynamo's trusted locations
+        /// </summary>
+        /// <param name="path">The path to be removed from the trusted locations</param>
+        /// <returns>The true if the path was removed and false otherwise</returns>
+        internal bool RemoveTrustedLocation(string path)
+        {
+            string location = ValidateTrustedLocation(path);
+            if (settings != null && trustedLocations.RemoveAll(x => x.Equals(location)) >= 0)
+            {
+                settings.TrustedLocations = trustedLocations;
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Overrides the entire trusted locations list with the input argument
+        /// </summary>
+        /// <param name="paths"></param>
+        internal void SetTrustedLocations(IEnumerable<string> paths)
+        {
+            trustedLocations.Clear();
+            foreach (var item in paths)
+            {
+                AddTrustedLocation(item);
+            }
+            settings.TrustedLocations = trustedLocations;
+        }
+
+        #region Public members
+        /// <summary>
+        /// The actual instance stored in the Singleton class
+        /// </summary>
+        public static TrustedLocationsManager Instance
+        {
+            get { return lazy.Value; }
+        }
+
+        /// <summary>
+        /// Checks is the path is considered valid for Dynamo's trusted locations.
+        /// An exception is thrown if the path is considered invalid.
+        /// A path is considered valid if the following conditions are true:
+        /// 1. Path is not null and not empty.
+        /// 2. Path is an absolute path (not relative).
+        /// 4. Path has valid characters.
+        /// 5. Path exists.
+        /// 6. Dynamo has read permissions to access the path.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns>A normalized and validated path</returns>
+        /// <exception cref="ArgumentException">Input argument is null or empty.</exception>
+        /// <exception cref="ArgumentException">Input argument is not an absolute path.</exception>
+        /// <exception cref="ArgumentException">Path directory does not exist</exception>
+        /// <exception cref="System.Security.SecurityException">Dynamo does not have the required permissions.</exception>
+        public string ValidateTrustedLocation(string path)
```

Comment:
I find the name of this method confusing - it does not validate that the location is trusted - correct? Instead it validates that it *could* be trusted?

If that is correct, what about something like ValidateLocation?
Also why is this public?

Reply:
no good reason for public.
Switching to internal ValidateLocation